File: csrc/hcsr04.c
/*
 * Copyright (c) 2020 Francesco Pio Squillante

 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#define ZERYNTH_PRINTF

#include "zerynth.h"

#define TicksPerMicros  (_system_frequency / 1000000)

/*
 * Sleeps for a given microseconds time
 *
 * micros:  Number of microseconds to wait
 */
void pseudoSleepForMicros(uint32_t micros){
    uint32_t startTicks = *vosTicks();
    
    while (*vosTicks() - startTicks < (TicksPerMicros * micros));
}

/*
 * Pseudo-ICU simulation using the processor thick to count the microseconds passed
 * 
 * pin:     Pin number of the ICU
 * mode:    Trigger pulse mode (1: high; 0: low)
 * timeout: maximum amount of microseconds to wait for the pulse
 * 
 * returns: The length of the ICU pulse in microseconds 
 */
uint32_t pseudoICU(int pin, int mode, uint32_t timeout){
    uint32_t startTicks = *vosTicks();
    
    while (*vosTicks() - startTicks < (TicksPerMicros * timeout))
        if(!!vhalPinRead(pin) != !!mode) //is the pulse finished?
            break;
    
    return (*vosTicks() - startTicks) / TicksPerMicros;
}

/*
 * Reads the distance from a HCSR04 device using the pseudo-ICU simulation
 * 
 * Parameters
 * ----------
 *  triggerPin : int
 *      The Pin number to which give the trigger pulse
 *  echoPin : int
 *      The Pin number from which retreive the ICU pulse
 */
err_t HCSR04_readDistanceRaw(int32_t nArgs, PObject *self, PObject **args, PObject **res) {
    int32_t triggerPin, echoPin;
    
    if (parse_py_args("ii", nArgs, args, &triggerPin, &echoPin) != 2)
        return ERR_TYPE_EXC;
    
    vosSysLock(); //For better performance

    vhalPinWrite(triggerPin, 0);
    vhalPinWrite(triggerPin, 1);
    pseudoSleepForMicros(10);
    vhalPinWrite(triggerPin, 0);
    
    uint32_t start_ticks = *vosTicks();
    
    while(!vhalPinRead(echoPin));
    
    *res = PSMALLINT_NEW(pseudoICU(echoPin, 1, 35000));
    
    vosSysUnlock();
    
    return ERR_OK;
}

File: hcsr04.py
# Copyright (c) 2020 helllayde
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

class hcsr04:
    """
        Class representing a HCSR04 device.
        If correctly initialized and plugged, the device is able to estimate the distance
        between its position and the first occluder object using a supersonic burst

        Attributes
        ----------
        trigger : int
            Tringger Pin number
        echo : int
            Echo pin number 
        
    """

    @c_native("HCSR04_readDistanceRaw", ["csrc/hcsr04.c"], [])
    def _getDistanceRaw(trig, echo):
        """
            (static wrapper function, should not be used. Use hcsr04.getDistanceRaw() instead)
            Get the distance as length in microseconds of the ICU pulse.
            This function is a wrapper from the C souce

            Parameters
            ----------
            trig : int
                Trigger Pin number
            echo : int
                Echo pin number 
        """
        pass
    
    def getDistanceRaw(self):
        """
            Get the distance from the device as length in microseconds of the ICU pulse
        """
        return hcsr04._getDistanceRaw(self.trigger, self.echo)
    
    def getDistanceCM(self):
        """
            Get the distance normalized in centimeters from the device
        """
        return self.getDistanceRaw() / 58
    
    def getDistanceINCH(self):
        """
            Get the distance normalized in inches from the device
        """
        return self.getDistanceRaw() / 148
    
    def __init__(self, trigger, echo):
        """
            Initialize a new HCSR04 device using the given pin numbers.
            When interrogated, the device will be triggered with a 10 microsecond-length
            pulse on the trigger pin and the device will respond with a pulse on the
            echo pin representing the length of the distance between the device and the
            first occluder

            Parameters
            ----------
            trigger : int
                Trigger Pin number
            echo : int
                Echo Pin number
        """
        self.trigger = trigger
        self.echo = echo
        
        pinMode(self.trigger, OUTPUT)
        pinMode(self.echo, INPUT)